File: service/src/test/java/no/nav/melosys/service/brev/DokumentNavnServiceTest.java
package no.nav.melosys.service.brev;

import java.util.stream.Stream;

import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.BehandlingTestBuilder;
import no.nav.melosys.domain.Lovvalgsperiode;
import no.nav.melosys.domain.brev.Mottaker;
import no.nav.melosys.domain.kodeverk.LovvalgBestemmelse;
import no.nav.melosys.domain.kodeverk.Mottakerroller;
import no.nav.melosys.service.LovvalgsperiodeService;
import no.nav.melosys.service.dokument.BrevmottakerService;
import no.nav.melosys.service.dokument.DokgenService;
import no.nav.melosys.service.dokument.brev.mapper.DokumentproduksjonsInfoMapper;
import no.nav.melosys.service.ftrl.medlemskapsperiode.MedlemskapsperiodeService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static no.nav.melosys.domain.kodeverk.Mottakerroller.*;
import static no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper.FØRSTEGANG;
import static no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper.NY_VURDERING;
import static no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.INNVILGELSE_FOLKETRYGDLOVEN;
import static no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.TRYGDEAVTALE_GB;
import static no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.trygdeavtale.Lovvalgsbestemmelser_trygdeavtale_gb.UK_ART6_1;
import static no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.trygdeavtale.Lovvalgsbestemmelser_trygdeavtale_gb.UK_ART8_2;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class DokumentNavnServiceTest {

    @Mock
    private BrevmottakerService brevmottakerService;
    @Mock
    private DokgenService dokgenService;
    @Mock
    private LovvalgsperiodeService lovvalgsperiodeService;
    @Mock
    private MedlemskapsperiodeService medlemskapsperiodeService;

    private final DokumentproduksjonsInfoMapper dokumentproduksjonsInfoMapper = new DokumentproduksjonsInfoMapper();

    private DokumentNavnService dokumentNavnService;

    @BeforeEach
    void setUp() {
        dokumentNavnService = new DokumentNavnService(brevmottakerService, dokgenService, lovvalgsperiodeService, medlemskapsperiodeService);

    }

    @Test
    void utledDokumentNavnForProduserbartdokumentOgMottakerrolle_ikkeStorbritannia_forventetTittel() {
        var behandling = lagBehandling();
        when(dokgenService.hentDokumentInfo(INNVILGELSE_FOLKETRYGDLOVEN)).thenReturn(dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(INNVILGELSE_FOLKETRYGDLOVEN));


        String dokumentNavn = dokumentNavnService.utledDokumentNavnForProduserbartdokumentOgMottakerrolle(behandling, INNVILGELSE_FOLKETRYGDLOVEN, BRUKER);


        assertThat(dokumentNavn).isEqualTo(INNVILGELSE_FOLKETRYGDLOVEN.getBeskrivelse());
    }

    @ParameterizedTest
    @MethodSource("testparametre")
    void utledDokumentNavnForProduserbartdokumentOgMottakerrolle_StorbritanniaInnvilgelseOgAttestMedUlikeParametre_korrektTittel(boolean skalHaAttest, boolean erNyVurdering, Mottaker mottaker, String forventetTittel) {
        var behandling = lagBehandling();
        behandling.setType(erNyVurdering ? NY_VURDERING : FØRSTEGANG);

        if (!mottaker.erUtenlandskMyndighet()) {
            when(lovvalgsperiodeService.hentLovvalgsperiode(anyLong())).thenReturn(lagLovvalsperiode(skalHaAttest ? UK_ART6_1 : UK_ART8_2));
        }
        when(brevmottakerService.avklarMottaker(TRYGDEAVTALE_GB, Mottaker.medRolle(mottaker.getRolle()), behandling)).thenReturn(mottaker);
        when(dokgenService.hentDokumentInfo(TRYGDEAVTALE_GB)).thenReturn(dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(TRYGDEAVTALE_GB));


        String dokumentNavn = dokumentNavnService.utledDokumentNavnForProduserbartdokumentOgMottakerrolle(behandling, TRYGDEAVTALE_GB, mottaker.getRolle());


        assertThat(dokumentNavn).isEqualTo(forventetTittel);
    }

    @ParameterizedTest
    @MethodSource("testparametre")
    void utledDokumentNavnForProduserbartdokumentOgMottaker_StorbritanniaInnvilgelseOgAttestMedUlikeParametre_korrektTittel(boolean skalHaAttest, boolean erNyVurdering, Mottaker mottaker, String forventetTittel) {
        var behandling = lagBehandling();
        behandling.setType(erNyVurdering ? NY_VURDERING : FØRSTEGANG);

        if (!mottaker.erUtenlandskMyndighet()) {
            when(lovvalgsperiodeService.hentLovvalgsperiode(anyLong())).thenReturn(lagLovvalsperiode(skalHaAttest ? UK_ART6_1 : UK_ART8_2));
        }
        when(dokgenService.hentDokumentInfo(TRYGDEAVTALE_GB)).thenReturn(dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(TRYGDEAVTALE_GB));


        String dokumentNavn = dokumentNavnService.utledDokumentNavnForProduserbartdokumentOgMottaker(behandling, TRYGDEAVTALE_GB, mottaker, "");


        assertThat(dokumentNavn).isEqualTo(forventetTittel);
    }

    @ParameterizedTest
    @MethodSource("testparametre")
    void utledDokumentNavn_StorbritanniaInnvilgelseOgAttestMedUlikeParametre_korrektTittel(boolean skalHaAttest, boolean erNyVurdering, Mottaker mottaker, String forventetTittel) {
        var behandling = lagBehandling();
        behandling.setType(erNyVurdering ? NY_VURDERING : FØRSTEGANG);

        if (!mottaker.erUtenlandskMyndighet()) {
            when(lovvalgsperiodeService.hentLovvalgsperiode(anyLong())).thenReturn(lagLovvalsperiode(skalHaAttest ? UK_ART6_1 : UK_ART8_2));
        }

        String dokumentNavn = dokumentNavnService.utledTittel(behandling, TRYGDEAVTALE_GB, dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(TRYGDEAVTALE_GB), mottaker, null, null, null);


        assertThat(dokumentNavn).isEqualTo(forventetTittel);
    }

    private static Stream<Arguments> testparametre() {
        return Stream.of(
            Arguments.of(true, false, lagMottaker(BRUKER, "1234", null, null), "Vedtak om medlemskap, Attest for medlemskap i folketrygden"),
            Arguments.of(false, false, lagMottaker(BRUKER, "1234", null, null), "Vedtak om medlemskap"),
            Arguments.of(true, false, lagMottaker(ARBEIDSGIVER, null, "1234", null), "Kopi av vedtak om medlemskap, Attest for medlemskap i folketrygden"),
            Arguments.of(false, false, lagMottaker(ARBEIDSGIVER, null, "1234", null), "Kopi av vedtak om medlemskap"),
            Arguments.of(true, false, lagMottaker(UTENLANDSK_TRYGDEMYNDIGHET, null, null, "1234"), "Attest for medlemskap i folketrygden"),

            Arguments.of(true, true, lagMottaker(BRUKER, "1234", null, null), "Vedtak om medlemskap, Attest for medlemskap i folketrygden - endring"),
            Arguments.of(false, true, lagMottaker(BRUKER, "1234", null, null), "Vedtak om medlemskap - endring"),
            Arguments.of(true, true, lagMottaker(ARBEIDSGIVER, null, "1234", null), "Kopi av vedtak om medlemskap, Attest for medlemskap i folketrygden - endring"),
            Arguments.of(false, true, lagMottaker(ARBEIDSGIVER, null, "1234", null), "Kopi av vedtak om medlemskap - endring"),
            Arguments.of(true, true, lagMottaker(UTENLANDSK_TRYGDEMYNDIGHET, null, null, "1234"), "Attest for medlemskap i folketrygden - endring")
        );
    }

    private Behandling lagBehandling() {
        var behandling = BehandlingTestBuilder.builderWithDefaults().build();
        behandling.setId(1L);
        return behandling;
    }

    private Lovvalgsperiode lagLovvalsperiode(LovvalgBestemmelse bestemmelse) {
        var lovvalgsperiode = new Lovvalgsperiode();
        lovvalgsperiode.setBestemmelse(bestemmelse);
        return lovvalgsperiode;
    }

    private static Mottaker lagMottaker(Mottakerroller rolle, String aktørID, String orgnr, String institusjonsID) {
        Mottaker mottaker = Mottaker.medRolle(rolle);
        mottaker.setAktørId(aktørID);
        mottaker.setOrgnr(orgnr);
        mottaker.setInstitusjonID(institusjonsID);
        return mottaker;
    }
}

File: service/src/test/kotlin/no/nav/melosys/service/brev/DokumentNavnServiceTest.kt
package no.nav.melosys.service.brev

import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.Lovvalgsperiode
import no.nav.melosys.domain.brev.Mottaker
import no.nav.melosys.domain.forTest
import no.nav.melosys.domain.kodeverk.LovvalgBestemmelse
import no.nav.melosys.domain.kodeverk.Mottakerroller
import no.nav.melosys.domain.kodeverk.Mottakerroller.*
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper.FØRSTEGANG
import no.nav.melosys.domain.kodeverk.behandlinger.Behandlingstyper.NY_VURDERING
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.INNVILGELSE_FOLKETRYGDLOVEN
import no.nav.melosys.domain.kodeverk.brev.Produserbaredokumenter.TRYGDEAVTALE_GB
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.trygdeavtale.Lovvalgsbestemmelser_trygdeavtale_gb.UK_ART6_1
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.trygdeavtale.Lovvalgsbestemmelser_trygdeavtale_gb.UK_ART8_2
import no.nav.melosys.service.LovvalgsperiodeService
import no.nav.melosys.service.dokument.BrevmottakerService
import no.nav.melosys.service.dokument.DokgenService
import no.nav.melosys.service.dokument.brev.mapper.DokumentproduksjonsInfoMapper
import no.nav.melosys.service.ftrl.medlemskapsperiode.MedlemskapsperiodeService
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource


@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@ExtendWith(MockKExtension::class)
class DokumentNavnServiceTest {

    @MockK
    private lateinit var brevmottakerService: BrevmottakerService

    @MockK
    private lateinit var dokgenService: DokgenService

    @MockK
    private lateinit var lovvalgsperiodeService: LovvalgsperiodeService

    @MockK
    private lateinit var medlemskapsperiodeService: MedlemskapsperiodeService

    private val dokumentproduksjonsInfoMapper = DokumentproduksjonsInfoMapper()

    private lateinit var dokumentNavnService: DokumentNavnService

    @BeforeEach
    fun setUp() {
        dokumentNavnService = DokumentNavnService(brevmottakerService, dokgenService, lovvalgsperiodeService, medlemskapsperiodeService)
    }

    @Test
    fun `utledDokumentNavnForProduserbartdokumentOgMottakerrolle ikke Storbritannia skal returnere forventet tittel`() {
        val behandling = lagBehandling()
        every { dokgenService.hentDokumentInfo(INNVILGELSE_FOLKETRYGDLOVEN) } returns dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(
            INNVILGELSE_FOLKETRYGDLOVEN
        )


        val dokumentNavn =
            dokumentNavnService.utledDokumentNavnForProduserbartdokumentOgMottakerrolle(behandling, INNVILGELSE_FOLKETRYGDLOVEN, BRUKER)


        dokumentNavn shouldBe INNVILGELSE_FOLKETRYGDLOVEN.beskrivelse
    }

    @ParameterizedTest
    @MethodSource("testparametre")
    fun `utledDokumentNavnForProduserbartdokumentOgMottakerrolle Storbritannia innvilgelse og attest med ulike parametre skal returnere korrekt tittel`(
        skalHaAttest: Boolean,
        erNyVurdering: Boolean,
        mottaker: Mottaker,
        forventetTittel: String
    ) {
        val behandling = lagBehandling()
        behandling.type = if (erNyVurdering) NY_VURDERING else FØRSTEGANG

        if (!mottaker.erUtenlandskMyndighet()) {
            every { lovvalgsperiodeService.hentLovvalgsperiode(any()) } returns lagLovvalsperiode(if (skalHaAttest) UK_ART6_1 else UK_ART8_2)
        }
        every { brevmottakerService.avklarMottaker(TRYGDEAVTALE_GB, Mottaker.medRolle(mottaker.rolle!!), behandling) } returns mottaker
        every { dokgenService.hentDokumentInfo(TRYGDEAVTALE_GB) } returns dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(TRYGDEAVTALE_GB)


        val dokumentNavn = dokumentNavnService.utledDokumentNavnForProduserbartdokumentOgMottakerrolle(behandling, TRYGDEAVTALE_GB, mottaker.rolle!!)


        dokumentNavn shouldBe forventetTittel
    }

    @ParameterizedTest
    @MethodSource("testparametre")
    fun `utledDokumentNavnForProduserbartdokumentOgMottaker Storbritannia innvilgelse og attest med ulike parametre skal returnere korrekt tittel`(
        skalHaAttest: Boolean,
        erNyVurdering: Boolean,
        mottaker: Mottaker,
        forventetTittel: String
    ) {
        val behandling = lagBehandling()
        behandling.type = if (erNyVurdering) NY_VURDERING else FØRSTEGANG

        if (!mottaker.erUtenlandskMyndighet()) {
            every { lovvalgsperiodeService.hentLovvalgsperiode(any()) } returns lagLovvalsperiode(if (skalHaAttest) UK_ART6_1 else UK_ART8_2)
        }
        every { dokgenService.hentDokumentInfo(TRYGDEAVTALE_GB) } returns dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(TRYGDEAVTALE_GB)


        val dokumentNavn = dokumentNavnService.utledDokumentNavnForProduserbartdokumentOgMottaker(behandling, TRYGDEAVTALE_GB, mottaker, "")


        dokumentNavn shouldBe forventetTittel
    }

    @ParameterizedTest
    @MethodSource("testparametre")
    fun `utledDokumentNavn Storbritannia innvilgelse og attest med ulike parametre skal returnere korrekt tittel`(
        skalHaAttest: Boolean,
        erNyVurdering: Boolean,
        mottaker: Mottaker,
        forventetTittel: String
    ) {
        val behandling = lagBehandling()
        behandling.type = if (erNyVurdering) NY_VURDERING else FØRSTEGANG

        if (!mottaker.erUtenlandskMyndighet()) {
            every { lovvalgsperiodeService.hentLovvalgsperiode(any()) } returns lagLovvalsperiode(if (skalHaAttest) UK_ART6_1 else UK_ART8_2)
        }


        val dokumentNavn = dokumentNavnService.utledTittel(
            behandling = behandling,
            produserbartdokument = TRYGDEAVTALE_GB,
            dokumentproduksjonsInfo = dokumentproduksjonsInfoMapper.hentDokumentproduksjonsInfo(TRYGDEAVTALE_GB),
            mottaker = mottaker,
            mottakerRolle = null,
            brevbestilling = null,
            standardTekst = null,
        )


        dokumentNavn shouldBe forventetTittel
    }

    private fun lagBehandling(): Behandling = Behandling.forTest {
        id = 1L
    }

    private fun lagLovvalsperiode(bestemmelse: LovvalgBestemmelse) = Lovvalgsperiode().apply {
        this.bestemmelse = bestemmelse
    }

    fun testparametre() =
        listOf(
            Arguments.of(true, false, lagMottaker(BRUKER, "1234", null, null), "Vedtak om medlemskap, Attest for medlemskap i folketrygden"),
            Arguments.of(false, false, lagMottaker(BRUKER, "1234", null, null), "Vedtak om medlemskap"),
            Arguments.of(
                true,
                false,
                lagMottaker(ARBEIDSGIVER, null, "1234", null),
                "Kopi av vedtak om medlemskap, Attest for medlemskap i folketrygden"
            ),
            Arguments.of(false, false, lagMottaker(ARBEIDSGIVER, null, "1234", null), "Kopi av vedtak om medlemskap"),
            Arguments.of(true, false, lagMottaker(UTENLANDSK_TRYGDEMYNDIGHET, null, null, "1234"), "Attest for medlemskap i folketrygden"),

            Arguments.of(
                true,
                true,
                lagMottaker(BRUKER, "1234", null, null),
                "Vedtak om medlemskap, Attest for medlemskap i folketrygden - endring"
            ),
            Arguments.of(false, true, lagMottaker(BRUKER, "1234", null, null), "Vedtak om medlemskap - endring"),
            Arguments.of(
                true,
                true,
                lagMottaker(ARBEIDSGIVER, null, "1234", null),
                "Kopi av vedtak om medlemskap, Attest for medlemskap i folketrygden - endring"
            ),
            Arguments.of(false, true, lagMottaker(ARBEIDSGIVER, null, "1234", null), "Kopi av vedtak om medlemskap - endring"),
            Arguments.of(
                true,
                true,
                lagMottaker(UTENLANDSK_TRYGDEMYNDIGHET, null, null, "1234"),
                "Attest for medlemskap i folketrygden - endring"
            )
        )

    private fun lagMottaker(rolle: Mottakerroller, aktørID: String?, orgnr: String?, institusjonsID: String?): Mottaker =
        Mottaker.medRolle(rolle).apply {
            this.aktørId = aktørID
            this.orgnr = orgnr
            this.institusjonID = institusjonsID
        }
}
